### DISCUSSION TITLE
Price per Trips

### DISCUSSION THREAD

**Discussion by LukeSkywalker993** [teslamate-org/teslamate#2433] [Ideas]
Hello,
Would it be possible to integrate a price per trip into the list under "Trips"?
Of course, this price should be calculated like the total ad after the selected period.

thank you.

**Comment by cwanja** [comment]
Costs are calculated based upon charging, so if you charge during a date range of the `Trip` dashboard it is shown:
![Screen Shot 2022-02-01 at 12 37 47 PM](https://user-images.githubusercontent.com/54645561/152030017-b0ccb282-fa11-4f4b-87ef-e9e404ef0d30.png)

**Reply by LukeSkywalker993** [reply]
Yes, but i like to seen the cost per drive here:
depending on the selected period and the charging costs in the period.
After "consumption" another line with "costs"

The costs of your picture are not for the single trip, but for the whole period.

![grafik](https://user-images.githubusercontent.com/59370744/152051959-b88329a2-9890-45d8-a178-71ebe3acf36f.png)

**Reply by cwanja** [reply]
So the ask here is to add a cost per drive then? Struggling to see the benefit. You would historically never look at the cost _per_ drive in any other car, unless it is for tax purposes. How are you going to calculate a cost per kW cost for a drive? 

This seems like an add-on potentially to the `Drives - Dutch Tax` dashboard that is in the Reports folder.

**Reply by LukeSkywalker993** [reply]
I find that very interesting. There are many ways to load for free or to load overpriced. There is also a monthly flat rate, which I have to convert manually to all flat charges at the end of the month (monthly price/charged kwh, and then add this to the charge by hand).
It would be in the right place under drives. That's correct.

So the charging costs are quite complex and I think in countries like Germany where you pay from 0.00 to 0.67€ per kwh it is interesting to be able to see the costs and the difference.

**Reply by cwanja** [reply]
How are you figuring out a cost per kW for a drive? For example, if you charge during two different "price points" - how are you going to divide the drive and used kW against the cost to charge? 

Maybe it is just my mind cannot wrap around the ask. Maybe it is just the difference between US / Germany. Based upon this statement:

> So the charging costs are quite complex and I think in countries like Germany where you pay from 0.00 to 0.67€ per kwh it is interesting to be able to see the costs and the difference.

It still seems like you are wanting to see the cost per _charging_ session.

**Reply by LukeSkywalker993** [reply]
In your picture, the cost for the period is 0.980.
And I would like to have divided these costs into "drives" on the route.

If I choose a period from 01/01/2022 to 01/31/2022, I see the costs for the period in trips.
You can't see what a single drive cost.

Let's say I paid for the first two weeks, but not for the second two weeks in January. Depending on how I choose the period, I can see what the trip cost through the charging costs.

I can see the cost per charging session.
If the price is fixed, I have set it, it works.

**Reply by LukeSkywalker993** [reply]
I would calculate it like this.
Cost from selected period = $cost_period

$cost_period / 100 km = $cost_per_KM

$Drive_cost = $cost_per_KM * $driven_km

$Drive_cost is what we need.

**Reply by cwanja** [reply]
I found that if you are on the `Statistics` dashboard and click a value in a row of the `Energy Charged` column, it will take you to the `Charging Stats` dashboard for that month and give you an average cost per 100 miles. I assume it would be km if your settings are configured for that. 
**Statistics**
![Screen Shot 2022-02-02 at 11 46 57 AM](https://user-images.githubusercontent.com/54645561/152208928-03df1360-2e52-4f40-a47c-0c3a77924f11.png)

**Charging Stats**
![Screen Shot 2022-02-02 at 11 46 34 AM](https://user-images.githubusercontent.com/54645561/152208738-e7c6dee8-9b9e-45bc-b6c1-58f56261cb84.png)

**Reply by LukeSkywalker993** [reply]
But that has nothing to do with what I would like.
In this view, you would have to manually calculate the cost of each trip.
"39.4 / 863 * km driven".
If I change the period, you have to calculate again.

A column would be great, where the costs are calculated. 

![grafik](https://user-images.githubusercontent.com/59370744/152221780-9b14bc0b-2fa5-4040-a865-ffa637c3a607.png)

**Comment by cschanzle** [comment]
I was curious about this same statistic - to know the cost per trip.  We (should) have the costs per charge and how much energy was added, and we know how much energy we have at the start of the trip.  For a given trip, we know the kWh used, so the only semi-difficult thing is to calculate the 'value' or cost of the energy in the battery, which could be done using a last in/first out method to add up our previous charges (kWh, $) until we get to our current charge level.

Let me use a contrived example.  Say I just took a drive that consumed 25 kWh and I want to know how much that cost me, but I've charged several times at different rates recently.

Before the drive in question, there was 50 kWh in the battery.  We need to calculate the cost of that 50 kWh.  Say my most recent charge was 20 kWh at $0.20/kWh ($4.00), so that's all in there.  Since 50-20 > 0, look at previous charge.  The 2nd most recent charge added 20 kWh at $0.60 ($12.00)...so our total cost for the last 40 kWh is  $4 + $12 = $16.  Since 50-40 = 10, which is > 0, again look at the next previous charge.  The 3rd most recent charge added 50 kWh at $0.20 ($10.00), but we have only 10 kWh of that charge left "in the tank," so 10/50 * 10 = $2.00.  Now we know the total cost of the 50 kWh stored in the battery before the trip:  $4 + $12 + $2 = $18.

Thus, my trip consuming 25 kWh of my 50 kWh valued at $18 is 25/50 * 18 = $9.

It's a bit convoluted, but programmatically straightforward.  I'm sure some people would find it "interesting."  And isn't processing all this data to discover "interesting" facts what it's all about?

Thanks for considering this!

**Comment by swiffer** [comment]
thanks sharing your example - i guess it will get more complicated the more drives and charges are happening within a trip.

I'm thinking about implementing something like this - a more basic approach has been done proposed here: https://github.com/teslamate-org/teslamate/pull/4678

(cost per km based on charging cost and driven distance - not factoring in soc at beginning and end of charge)

i suggested considering soc at beginning and ending of trip here, that's what i would open a PR for tomorrow: https://github.com/teslamate-org/teslamate/pull/4678#issuecomment-2830284946

still not fully fifo but as long as at least one charge happened within the period of the trip it would give some reasonable numbers.

@cschanzle - what do you think - would that be sufficient?

once we have average cost per km within the trip it's easy to calculate each drives cost.

**Comment by o6uoq** [comment]
Was the price-per-trip feature ever added? My dashboard shows cost in the trip view as being nil.

**Reply by JakobLichterfeld** [reply]
Yes, it shows n/a if no charge happend in the period

**Reply by swiffer** [reply]
as Jacob said it's showing the cost for charging sessions that occured in the period of the trip so far.

planned some improvements but haven't managed to do that so far - PR #4678 is still in draft.

**Reply by o6uoq** [reply]
I can see the cost for charging sessions but what I am looking for is cost per trip. This does not seem to be working for me, even tho I have the charging costs linked up and working.
